test/widgets/compose_box_test.dart

```diff
@@ -385,97 +407,251 @@ void main() {
+          checkComposeBox(isShown: true);
+        });
       });
 
-      testWidgets('deactivated user becomes active -> '
-          'banner is replaced with the compose box', (tester) async {
-        final deactivatedUser = eg.user(isActive: false);
-        await prepareComposeBox(tester, narrow: dmNarrowWith(deactivatedUser),
-          users: [deactivatedUser]);
-        checkComposeBox(isShown: false);
+      group('group DMs', () {
+        testWidgets('compose box replaced with a banner', (tester) async {
+          final deactivatedUsers = [eg.user(isActive: false), eg.user(isActive: false)];
+          await prepareComposeBox(tester, narrow: groupDmNarrowWith(deactivatedUsers),
+            users: deactivatedUsers);
+          checkComposeBox(isShown: false);
+        });
 
-        await changeUserStatus(tester, user: deactivatedUser, isActive: true);
-        checkComposeBox(isShown: true);
+        testWidgets('at least one user becomes deactivated -> '
+            'compose box is replaced with a banner', (tester) async {
+          final activeUsers = [eg.user(isActive: true), eg.user(isActive: true)];
+          await prepareComposeBox(tester, narrow: groupDmNarrowWith(activeUsers),
+            users: activeUsers);
+          checkComposeBox(isShown: true);
+
+          await changeUserStatus(tester, user: activeUsers[0], isActive: false);
+          checkComposeBox(isShown: false);
+        });
+
+        testWidgets('all deactivated users become active -> '
+            'banner is replaced with the compose box', (tester) async {
+          final deactivatedUsers = [eg.user(isActive: false), eg.user(isActive: false)];
+          await prepareComposeBox(tester, narrow: groupDmNarrowWith(deactivatedUsers),
+            users: deactivatedUsers);
+          checkComposeBox(isShown: false);
+
+          await changeUserStatus(tester, user: deactivatedUsers[0], isActive: true);
+          checkComposeBox(isShown: false);
+
+          await changeUserStatus(tester, user: deactivatedUsers[1], isActive: true);
+          checkComposeBox(isShown: true);
+        });
       });
     });
 
-    group('group DMs', () {
-      testWidgets('compose box replaced with a banner', (tester) async {
-        final deactivatedUsers = [eg.user(isActive: false), eg.user(isActive: false)];
-        await prepareComposeBox(tester, narrow: groupDmNarrowWith(deactivatedUsers),
-          users: deactivatedUsers);
-        checkComposeBox(isShown: false);
+    group('in topic/channel narrow according to channel post policy', () {
+      void checkComposeBox({required bool isShown}) => checkComposeBoxIsShown(isShown,
+        bannerLabel: zulipLocalizations.errorBannerCannotPostInChannelLabel);
+
+      final testCases = [
+        (ChannelPostPolicy.unknown,        UserRole.unknown,       true),
+        (ChannelPostPolicy.unknown,        UserRole.guest,         true),
+        (ChannelPostPolicy.unknown,        UserRole.member,        true),
```

Comment:
These can be adapted to produce those unit tests that `hasPostingPermission` should get.

Then once that has unit tests, the tests for this UI code can be simpler: they only need to exercise the cases that are different as far as the UI code sees things, not cases that differ only within the implementation of the model method. So e.g. cases for a channel narrow and a topic narrow, and needing a banner and not needing one; but not for all the nuances of when one is needed and when it isn't.